src/hcpp-hpt.c

```diff
@@ -275,12 +304,163 @@ inline void init_hc_deque_t(hc_deque_t * hcdeq, place_t * pl){
+            }
+        }
+#else
+        HASSERT(flags == NONE);
+#endif
+        return malloc(nbytes);
+#ifdef HC_CUDA
+    } else if (is_nvgpu_place(pl)) {
+        HASSERT(flags == NONE);
+        void *ptr;
+        HASSERT(pl->cuda_id >= 0);
+        CHECK_CUDA(cudaSetDevice(pl->cuda_id));
+        const cudaError_t alloc_err = cudaMalloc((void **)&ptr, nbytes);
+        if (alloc_err != cudaSuccess) {
+#ifdef VERBOSE
+            fprintf(stderr, "Allocation at NVGPU place failed with reason "
+                    "\"%s\"\n", cudaGetErrorString(alloc_err));
+#endif
+            return NULL;
+        } else {
+            return ptr;
+        }
+#endif
+    } else {
+        unsupported_place_type_err(pl);
+        return NULL; // will never reach here
+    }
 }
 
+int is_pinned_cpu_mem(void *ptr) {
+    return hclib_memory_tree_contains(ptr, &hcpp_context->pinned_host_allocs);
+}
+
+void hclib_free_at(place_t *pl, void *ptr) {
+    if (is_cpu_place(pl)) {
+        if (is_pinned_cpu_mem(ptr)) {
+            hclib_memory_tree_remove(ptr, &hcpp_context->pinned_host_allocs);
+            CHECK_CUDA(cudaFreeHost(ptr));
+        } else {
+            free(ptr);
+        }
+#ifdef HC_CUDA
+    } else if (is_nvgpu_place(pl)) {
+        CHECK_CUDA(cudaFree(ptr));
+#endif
+    } else {
+        unsupported_place_type_err(pl);
+    }
+}
+
+// Used to wrap the creation of an async copy task that is dependent on some DDFs
+static void async_gpu_task_launcher(void *arg) {
+    gpu_task_t *task = (gpu_task_t *)arg;
+    spawn_gpu_task((task_t *)task);
+}
+
+/*
+ * TODO Currently doesn't support await on other DDFs, nor do communication
+ * tasks
+ */
+hclib_ddf_t *hclib_async_copy(place_t *dst_pl, void *dst, place_t *src_pl,
+        void *src, size_t nbytes, hclib_ddf_t **ddf_list, void *user_arg) {
+    gpu_task_t *task = malloc(sizeof(gpu_task_t));
+    task->t._fp = NULL;
+    task->t.is_asyncAnyType = 0;
+    task->t.ddf_list = NULL;
+    task->t.args = NULL;
+    task->t.place = NULL;
+
+    hclib_ddf_t *ddf = hclib_ddf_create();
+    task->gpu_type = GPU_COMM_TASK;
+    task->ddf_to_put = ddf;
+    task->arg_to_put = user_arg;
+
+    task->gpu_task_def.comm_task.src_pl = src_pl;
+    task->gpu_task_def.comm_task.dst_pl = dst_pl;
+    task->gpu_task_def.comm_task.src = src;
+    task->gpu_task_def.comm_task.dst = dst;
+    task->gpu_task_def.comm_task.nbytes = nbytes;
+
+#ifdef VERBOSE
+    fprintf(stderr, "hclib_async_copy: dst_pl=%p dst=%p src_pl=%p src=%p "
+            "nbytes=%lu ddf_list=%p\n", dst_pl, dst, src_pl, src, nbytes,
+            ddf_list);
+#endif
+
+    if (ddf_list) {
+        hclib_async(async_gpu_task_launcher, task, ddf_list, NULL, NULL, 0);
+    } else {
+        spawn_gpu_task((task_t *)task);
+    }
+
+    return ddf;
+}
+
+hclib_ddf_t *hclib_async_memset(place_t *pl, void *ptr, int val, size_t nbytes,
+        hclib_ddf_t **ddf_list, void *user_arg) {
+    gpu_task_t *task = malloc(sizeof(gpu_task_t));
```

Comment:
Same comment. Single word comments in code reviews make it hard for people to fix problems in the code, because they have no way to understand the reasoning behind them.